tests/quality_pragmaitc/test_refinement.py

```diff
@@ -24,63 +30,108 @@ def test_shouldBeApplicable():
+def test_shouldBeNotApplicable():
+    goal = Goal()
+    task = Task()
+    delegation = Delegation()
+
+    context = Context("C1")
 
-        task.addApplicableContext(context)
-        #goal.addApplicableContext(context)
-        delegation.addApplicableContext(context)
+    task.addApplicableContext(context)
 
-        wrongContext = Context("C2")
-        fullContext = []
-        fullContext.add(wrongContext)
+    goal.addApplicableContext(context)
 
-        #assert False == goal.isApplicable(fullContext)
-        assert False == task.isApplicable(fullContext)
-        assert False == delegation.isApplicable(fullContext)
+    delegation.addApplicableContext(context)
 
-    def test_aTaskShouldBeAchievable():
-        commonMetrics = CommonMetrics()
-        task = Task()
+    wrongContext = Context("C2")
+    fullContext = []
+    fullContext.append(wrongContext)
 
-        currentContext = Context("C1")
-        fullContext = []
-        fullContext.append(currentContext)
+    assert False is goal.isApplicable(fullContext)
+    assert False is task.isApplicable(fullContext)
+    assert False is delegation.isApplicable(fullContext)
 
-        qc = QualityConstraint(currentContext, commonMetrics.SECONDS, 15, 'LESS_OR_EQUAL_TO')
 
-        task.addApplicableContext(currentContext)
-        task.setProvidedQuality(currentContext, commonMetrics.SECONDS, 12)
+def test_aTaskShouldBeAchievable():
+    task = Task()
 
-        interp = Interpretation()
-        interp.addQualityConstraint(qc)
+    currentContext = Context("C1")
+    fullContext = []
+    fullContext.append(currentContext)
 
-        assert True == task in task.isAchievable(fullContext, interp).getTasks()
+    qc = QualityConstraint(
+        currentContext, CommonMetrics.SECONDS, 15, 'LESS_OR_EQUAL_TO')
+
+    task.addApplicableContext(currentContext)
+    task.setProvidedQuality(currentContext, CommonMetrics.SECONDS, 12)
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    assert True is task in task.isAchievable(
+        fullContext, interp).getTasks()
+
+
+def test_shouldAddSeveralContextsAtOnce():
+    context1 = Context("C1")
+    context2 = Context("C2")
+
+    task = Task()
+    originalSize = len(task.getApplicableContext())
+    set = []
 
-    def test_shouldAddSeveralContextsAtOnce():
-        context1 = Context("C1")
-        context2 = Context("C2")
+    set.append(context1)
+    set.append(context2)
+
+    task.addApplicableContext(set)
+
+    assert 2 == len(task.getApplicableContext()) - originalSize
+
+
+def aNonApplicableRootGoalIsNotAchievable():
+    goal = Goal(False)
+    current = Context("C1")
+    fullContext = []
+
+    qc = QualityConstraint(current, CommonMetrics.SECONDS,
+                           15, Comparison.LESS_OR_EQUAL_TO)
+    goal.addApplicableContext(Context("C2"))
+
+    interp = Interpretation()
+    interp.addQualityConstraint(qc)
+
+    assert goal.isAchievable(fullContext, interp) is None
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/lesunb/p2psimpy/pullRequest?prid=4861978)